certora/specs/bridge.spec

```diff
@@ -0,0 +1,526 @@
+}
+
+// Checks that it isn't possible to gain from transforming dynamic to static
+// and back.
+// This is violated because the mul and div are not inverses of each other,
+// therefore can lead to mul(div(a,b),b) > a (depends on remainder value).
+rule dynamicToStaticInversible2(uint256 amount)
+{
+    // We assume both indexes (L1,L2) are represented in Ray (1e27).
+    address asset;
+    requireRayIndex(asset);
+    // Just for debugging.
+    uint256 indexL1 = LENDINGPOOL_L1.liquidityIndexByAsset(asset);
+    uint256 stat = _dynamicToStaticAmount_Wrapper(amount, asset, LENDINGPOOL_L1);
+    uint256 dynm = _staticToDynamicAmount_Wrapper(stat, asset, LENDINGPOOL_L1);
+    assert amount == dynm;
+    assert amount >= dynm;
+}
+
+// We make sure that the message sent booleans are always false,
+// meaning that, according to our implementation, no external call can
+// invoke receiveRewards and withdraw in Bridge.sol, but only through the 
+// designated functions in the harnessed Bridge contract.
+invariant alwaysUnSent(env e)
+   !withdrawMessageStatus(e) && !bridgeRewardsMessageStatus(e)
+
+// Check consistency of 'asset' being registered as the underlying
+// token of 'AToken', both in the AToken contract, and also in the 
+// mapping _aTokenData.
+// We exclude the 'initialize' function since it is called only once
+// in the code. 
+invariant underlying2ATokenConsistency(address AToken, address asset)
+     (asset !=0 <=> AToken !=0) 
+     =>
+     (getUnderlyingAssetHelper(AToken) == asset 
+     <=>
+     getUnderlyingAssetOfAToken(AToken) == asset)
+     filtered{f-> excludeInitialize(f)}
+
+// Check consistency of 'asset' being registered as the underlying
+// token of 'AToken', and 'AToken' connected to 'asset' in the lending pool.
+// We exclude the 'initialize' function since it is called only once
+// in the code. 
+invariant ATokenAssetPair(address asset, address AToken)
+    (asset !=0 <=> AToken !=0) 
+    =>
+    (getUnderlyingAssetHelper(AToken) == asset 
+    <=>
+    getATokenOfUnderlyingAsset(LENDINGPOOL_L1, asset) == AToken)
+    filtered{f -> excludeInitialize(f)}
+
+// The aToken-asset pair should be correctly registered after calling
+// initialize, right after the constructor.
+// This is complementary to the two invariants above.
+rule initializeIntegrity(address AToken, address asset)
+{
+    env e;
+    calldataarg args;
+
+    // Post-constructor conditions
+    require getUnderlyingAssetHelper(AToken) == 0;
+    require getATokenOfUnderlyingAsset(LENDINGPOOL_L1, asset) == 0;
+    
+    initialize(e, args);
+
+    assert (asset !=0 && AToken !=0) => (
+        (getUnderlyingAssetHelper(AToken) == asset 
+        <=>
+        getUnderlyingAssetHelper(AToken) == asset)
+     &&
+        (getUnderlyingAssetHelper(AToken) == asset 
+        <=>
+        getATokenOfUnderlyingAsset(LENDINGPOOL_L1, asset) == AToken));
+}
+
+rule sanity(method f) {
+    env e;
+    calldataarg args;
+    f(e, args);
+    assert false;
```

Comment:
Better to make this rule with the calls to setup function and the use callFunctionSetParams,  then we check that we don't have sanity issues due to the setup

Reply:
So you mean check reachability of 'callFunctionSetParams' ?
I can write it as a rule, though I already did.
Unreachability is obtained for the withdraw and receiveRewards functions, but this is intended, as to not let the user call those functions from L1 - this is a result of the _alwaysUnSent(e)_ entrancy guard.

Reply:
So you mean check reachability of 'callFunctionSetParams' ?
I can write it as a rule, though I already did.
Unreachability is obtained for the withdraw and receiveRewards functions, but this is intended, as to not let the user call those functions from L1 - this is a result of the _alwaysUnSent(e)_ entrancy guard.

Reply:
So you mean check reachability of 'callFunctionSetParams' ?
I can write it as a rule, though I already did.
Unreachability is obtained for the withdraw and receiveRewards functions, but this is intended, as to not let the user call those functions from L1 - this is a result of the _alwaysUnSent(e)_ entrancy guard.

Reply:
So you mean check reachability of 'callFunctionSetParams' ?
I can write it as a rule, though I already did.
Unreachability is obtained for the withdraw and receiveRewards functions, but this is intended, as to not let the user call those functions from L1 - this is a result of the _alwaysUnSent(e)_ entrancy guard.

Reply:
in general, instead of f(e,args) use the callFunctionSetParams, and call the setup functions. they way you are more certain that you setup is ok. this will be good for others that might add more require to the setup